### PR TITLE
Introduce meta eval metric for general NLG tasks

### PR DESCRIPTION
**Pull request by pfliu-nlp**
Previously, we have meta-evaluation metrics (and tasks) specifically defined for WMT, which are too dataset dependent and not generalized enough for other NLG tasks, such as summarization and data-to-text. This PR aims to introduce meta-evaluation metrics for generation tasks such as summarization and data-to-text.

### CODE REVIEW

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 from typing import Optional, Union
 
 import numpy as np
+from scipy import stats
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:9]: I think L10 is not necessary anymore.

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 2]
```diff
@@ -18,6 +19,142 @@
...
 
 
+@dataclass
+@common_registry.register("CorrelationNLGConfig")
+class CorrelationNLGConfig(MetricConfig):
+    """Configuration of a correlation for general NLG tasks.
+
+    Args:
+        level: there are following levels: sample level, system level and dataset level
+                See more details: https://aclanthology.org/2020.emnlp-main.751.pdf
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:29]: I think it is better to be renamed this to *strategy* as referred in the section 2.5 of the paper. "level" seems somewhat confusing because we have already AnalysisLevels.

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 3]
```diff
@@ -18,6 +19,142 @@
...
+        func_name: the method for calculating the correlation (e.g., Spearman)
+    """
+
+    level: str = "sample"
+    func_name: str = "spearmanr"
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:34]: where these defaults come from?

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 4]
```diff
@@ -18,6 +19,142 @@
...
 
+@dataclass
+@common_registry.register("CorrelationNLGConfig")
+class CorrelationNLGConfig(MetricConfig):
+    """Configuration of a correlation for general NLG tasks.
+
+    Args:
+        level: there are following levels: sample level, system level and dataset level
+                See more details: https://aclanthology.org/2020.emnlp-main.751.pdf
+        func_name: the method for calculating the correlation (e.g., Spearman)
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:30]: - "func_name" does not tell what this value is. Some other names such as "correlation" is suitable.
- You also need to add the full list of applicable values for this member. The example "Spearman" is not feasible.

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 5]
```diff
@@ -18,6 +19,142 @@
...
+
+    def to_metric(self) -> Metric:
+        """See MetricConfig.to_metric."""
+        return CorrelationNLG(self)
+
+    def get_correlation_func(self, name: str):
+        """Get correlation function based on function name.
+
+        TODO(pengfei): organize this in a better way
+
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:44]: Leave this TODO outside the docstring.

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 6]
```diff
@@ -18,6 +19,142 @@
...
+            return stats.kendalltau
+        else:
+            raise ValueError(f"The correlation function {name} hasn't been supported")
+
+
+class CorrelationNLG(Metric):
+    """A metric that calculates correlations."""
+
+    n_samples = 0
+    n_systems = 0
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:62]: These values should be in the stats, not object members in this class

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 7]
```diff
@@ -18,6 +19,142 @@
...
+        return False
+
+    def uses_customized_aggregate(self) -> bool:
+        """See Metric.uses_customized_aggregate."""
+        return True
+
+    def calc_stats_from_data(
+        self,
+        true_data: list[list[float]],
+        pred_data: list[list[float]],
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:75]: These values are typed as `list[Any]`

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 8]
```diff
@@ -18,6 +19,142 @@
...
+        elif config.level == "system":
+            return SimpleMetricStats(
+                np.array([true + pred for true, pred in zip(true_data, pred_data)])
+            )
+        else:  # TODO(Pengfei): better way to organize different levels
+            return SimpleMetricStats(
+                np.array(
+                    [[true[0], pred[0]] for true, pred in zip(true_data, pred_data)]
+                )
+            )
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:101]: As far as I read this class, there are only dataset/system/sample choices for `level`. The class has to implement a full list of conditions for these values, and raise errors for any other values.

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 9]
```diff
@@ -18,6 +19,142 @@
...
+    def _aggregate_stats(self, stats: MetricStats) -> np.ndarray:
+        """See Metric.aggregate_stats."""
+        if stats.is_batched():
+            data = stats.get_batch_data()
+            return data.reshape((data.shape[0], data.shape[-2] * data.shape[-1]))
+        else:
+            data = stats.get_data()
+            return data.reshape((data.shape[-2] * data.shape[-1]))
+
+    def calc_metric_from_aggregate_single(self, single_stat: np.ndarray) -> float:
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:112]: This should be a private function.

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 10]
```diff
@@ -18,6 +19,142 @@
...
+        """Calculate an aggregate correlation metric from a single segment or system.
+
+        Args:
+            single_stat: The stats for the single segment or system
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:116]: If the function takes/returns ndarrays, please describe applicable shapes as well.

File: explainaboard/metrics/nlg_meta_evaluation.py [hunk 11]
```diff
@@ -18,6 +19,142 @@
...
+        config = narrow(CorrelationNLGConfig, self.config)
+        corr_func = config.get_correlation_func(config.func_name)
+        if config.level == "dataset":
+            val = corr_func(single_stat[:, 0], single_stat[0:, 1])[0]
+        elif config.level == "sample":
+            val = np.mean(single_stat)
+        elif config.level == "system":
+            true_scores = np.sum(single_stat[:, 0 : int(self.n_systems)], axis=0)
+            pred_scores = np.sum(single_stat[:, int(self.n_systems) :], axis=0)
+            val = corr_func(true_scores, pred_scores)[0]
```
Comment (odashi) [inline explainaboard/metrics/nlg_meta_evaluation.py:131]: add the else clause to raise errors for any other values of `level`.

File: integration_tests/nlg_meta_eval_test.py [hunk 12]
```diff
@@ -39,3 +40,94 @@ def test_da_cs_en(self):
 
         self.assertGreater(len(sys_info.results.analyses), 0)
         self.assertGreater(len(sys_info.results.overall), 0)
+
+
+class MetaEvalNLGTest(unittest.TestCase):
```
Comment (odashi) [inline integration_tests/nlg_meta_eval_test.py:45]: Add tests to check that specifying invalid values of either `level` or `func_name` causes an error.

File: integration_tests/nlg_meta_eval_test.py [hunk 13]
```diff
@@ -39,3 +40,94 @@ def test_da_cs_en(self):
...
+    true_data = [[1, 2, 3, 4, 5], [2, 1, 4, 5, 2], [5, 4, 3, 2, 1]]
+    pred_data = [[2, 1, 3, 4, 5], [2, 4, 5, 5, 2], [5, 3, 4, 2, 1]]
+
+    def test_sample_level_spearmanr(self):
```
Comment (odashi) [inline integration_tests/nlg_meta_eval_test.py:49]: Add `-> None` annotation to all test functions so that the type checker captures these functions.

```suggestion
    def test_sample_level_spearmanr(self) -> None:
```

File: integration_tests/nlg_meta_eval_test.py [hunk 14]
```diff
@@ -39,3 +40,94 @@ def test_da_cs_en(self):
...
+
```
Comment (odashi) [inline integration_tests/nlg_meta_eval_test.py:50]: Inconsistent blank lines for several tests. I think these are not necessary to be added.

File: integration_tests/nlg_meta_eval_test.py [hunk 15]
```diff
@@ -39,3 +40,94 @@ def test_da_cs_en(self):
...
+        nlg_corr_config = CorrelationNLGConfig(level="sample", func_name="spearmanr")
+        corr_metric = nlg_corr_config.to_metric()
+        stats = corr_metric.calc_stats_from_data(self.true_data, self.pred_data)
+        stats_arr = corr_metric.aggregate_stats(stats)
+        val = corr_metric.calc_metric_from_aggregate_single(stats_arr)
+        self.assertAlmostEqual(val, 0.8162952, 3)
+
+        ci = corr_metric.calc_confidence_interval(stats, 0.05)
+        self.assertGreater(val, ci[0])
+        self.assertGreater(ci[1], val)
```
Comment (odashi) [inline integration_tests/nlg_meta_eval_test.py:120]: This is weak to capture bugs. Add some checks of the range of actual values as well.

File: integration_tests/nlg_meta_eval_test.py [hunk 16]
```diff
@@ -39,3 +40,94 @@ def test_da_cs_en(self):
...
+        nlg_corr_config = CorrelationNLGConfig(level="system", func_name="spearmanr")
+        corr_metric = nlg_corr_config.to_metric()
+        stats = corr_metric.calc_stats_from_data(self.true_data, self.pred_data)
+        stats_arr = corr_metric.aggregate_stats(stats)
+        val = corr_metric.calc_metric_from_aggregate(stats_arr)
+        self.assertAlmostEqual(val, 0.815789, 3)
+
+        ci = corr_metric.calc_confidence_interval(stats, 0.05)
+        self.assertGreater(val, ci[0])
+        self.assertGreater(ci[1], val)
```
Comment (odashi) [inline integration_tests/nlg_meta_eval_test.py:133]: ditto

File: integration_tests/nlg_meta_eval_test.py [hunk 17]
```diff
@@ -39,3 +40,94 @@ def test_da_cs_en(self):
...
+
+        nlg_corr_config = CorrelationNLGConfig(level="dataset", func_name="spearmanr")
+        corr_metric = nlg_corr_config.to_metric()
+        stats = corr_metric.calc_stats_from_data(true_data, pred_data)
+        stats_arr = corr_metric.aggregate_stats(stats)
+        val = corr_metric.calc_metric_from_aggregate(stats_arr)
+        self.assertAlmostEqual(val, 1.0, 3)
+
+
+class MetaEvalNLGCITest(unittest.TestCase):
```
Comment (odashi) [inline integration_tests/nlg_meta_eval_test.py:105]: It looks there is no tests for `level=dataset`

File: explainaboard/metrics/nlg_meta_evaluation.py (hunk 2 above)
Comment (pfliu-nlp) [reply explainaboard/metrics/nlg_meta_evaluation.py:29]: Sounds good. I will follow this suggestion.

File: explainaboard/metrics/nlg_meta_evaluation.py (hunk 4 above)
Comment (pfliu-nlp) [reply explainaboard/metrics/nlg_meta_evaluation.py:30]: sounds good

File: integration_tests/nlg_meta_eval_test.py (hunk 17 above)
Comment (pfliu-nlp) [reply integration_tests/nlg_meta_eval_test.py:105]: there indeed is one: `test_dataset_level_spearmanr`

File: explainaboard/metrics/nlg_meta_evaluation.py (hunk 3 above)
Comment (pfliu-nlp) [reply explainaboard/metrics/nlg_meta_evaluation.py:34]: I finally renamed `level` as `group_by` to keep consistent with how `meta_evaluation_wmt_da` task names it.

File: integration_tests/nlg_meta_eval_test.py (hunk 17 above)
Comment (odashi) [reply integration_tests/nlg_meta_eval_test.py:105]: It looks there are separate test classes `MetaEval{WMTDA/NLG/NLGCI}Test`. If they are testing different aspect of the functionality, you basically need to check all edge cases into every test. `test_dataset_level_spearmanr` is contained only in `MetaEvalNLGTest`

**Comment by pfliu-nlp** [comment]
@odashi, I should fix most of the comments. One remaining one could be fixed after we have a consensus (in PR https://github.com/neulab/ExplainaBoard/pull/527) about how to store the number of samples.

File: integration_tests/nlg_meta_eval_test.py (hunk 17 above)
Comment (pfliu-nlp) [reply integration_tests/nlg_meta_eval_test.py]: Sounds good. I can add one in `MetaEvalNLGCITest`.

**Comment by neubig** [comment]
#528 is merged, so I think this can be revisited

File: explainaboard/metrics/nlg_meta_evaluation.py (hunk 6 above)
Comment (odashi) [reply explainaboard/metrics/nlg_meta_evaluation.py:63]: This comment is not reflected.

File: integration_tests/meta_eval_wmt_da_test.py [hunk 18]
```diff
@@ -39,94 +43,170 @@ def test_da_cs_en(self):
...
+            ValueError, r"^The correlation " r"function illegal hasn't been supported"
+        ):
+            nlg_corr_config = CorrelationNLGConfig(
+                group_by="sample", correlation_type="illegal"
+            )
+            corr_metric = narrow(CorrelationNLG, nlg_corr_config.to_metric())
+            stats = corr_metric.calc_stats_from_data(self.true_data, self.pred_data)
+            stats_arr = corr_metric.aggregate_stats(stats)
+            val = corr_metric._calc_metric_from_aggregate_single(stats_arr)
+            self.assertAlmostEqual(val, 0.8162952, 3)
```
Comment (odashi) [inline integration_tests/meta_eval_wmt_da_test.py:68]: tip: If the test uses `assertRaises`, it is recommended to wrap only one statement to check if a particular statement should raise the error and any other statements can be processed appropriately.

There are two positions that the targeted error will occur: `calc_stats_from_data` and `aggregate_stats`. They should be tested separately.

```suggestion
    def test_illegal_correlation_type_calc_stats_from_data(self) -> None:
        with self.assertRaisesRegex(ValueError, r"^The correlation function"):
        nlg_corr_config = CorrelationNLGConfig(
            group_by="sample", correlation_type="illegal"
        )
        corr_metric = narrow(CorrelationNLG, nlg_corr_config.to_metric())
        
        with self.assertRaisesRegex(ValueError, r"^The correlation function"):
            corr_metric.calc_stats_from_data(self.true_data, self.pred_data)
            
    def test_illegal_correlation_type_aggregate_stats(self) -> None:
        with self.assertRaisesRegex(ValueError, r"^The correlation function"):
        nlg_corr_config = CorrelationNLGConfig(
            group_by="sample", correlation_type="illegal"
        )
        corr_metric = narrow(CorrelationNLG, nlg_corr_config.to_metric())
        stats = SimpleMetricStats(np.zeros(...))
        
        with self.assertRaisesRegex(ValueError, r"^The correlation function"):
            corr_metric.aggregate_stats(stats)
```

File: integration_tests/meta_eval_wmt_da_test.py [hunk 19]
```diff
@@ -39,94 +43,170 @@ def test_da_cs_en(self):
...
+
+    def test_illegal_group_type(self) -> None:
```
Comment (odashi) [inline integration_tests/meta_eval_wmt_da_test.py:70]: ditto

**Review by odashi — COMMENTED** [review]
Some previous comments are not yet reflected

**Review by odashi — APPROVED** [review]
